pkg/vm/engine/tae/logtail/service/server.go

```diff
@@ -0,0 +1,465 @@
+// Copyright 2021 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/fagongzi/goetty/v2"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+
+	"github.com/matrixorigin/matrixone/pkg/common/moerr"
+	"github.com/matrixorigin/matrixone/pkg/common/morpc"
+	"github.com/matrixorigin/matrixone/pkg/logutil"
+	"github.com/matrixorigin/matrixone/pkg/pb/api"
+	"github.com/matrixorigin/matrixone/pkg/pb/logtail"
+	"github.com/matrixorigin/matrixone/pkg/pb/timestamp"
+	"github.com/matrixorigin/matrixone/pkg/util/trace"
+)
+
+const (
+	LogtailServiceRPCName = "logtail-rpc"
+)
+
+// Timestamp determines current timestamp.
+type Timestamp interface {
```

Comment:
Can't define a clock yourself. Please use the clock in runtime

Reply:
It's a restrictive interface with an unreasonable name.
In the following commit for this PR, the name and method of this interface has been updated according to clock in runtime.